README
README
------

This is SDL_bgi, a graphics library based on SDL2. SDL_bgi provides
graphics primitives, and is much easier to use than SDL2.

SDL_bgi is nearly fully compatible with BGI, the Borland Graphics
Interface that was the de facto standard in PC graphics back in DOS
days. (It compiles and runs the original 'bgidemo.c' almost
perfectly.) It's pretty fast, very easy to use, and it also provides
extensions for ARGB colours and mouse support. It should run on any
platform supported by SDL2; it was only tested in Linux, though.

I wrote SDL_bgi because I wanted a simple but fast graphics library
for my experiments with fractals and cellular automata, using the BGI
syntax I'm used to. Obviously, native SDL2 functions can be used in
SDL_bgi programs.

A few examples are provided in the test/ directory.

If you don't need BGI compatibility, I suggest that you check out
SDL_gfx, a graphics library more complete and more powerful than
SDL_bgi.

Links:

0. The SDL library:
   https://www.libsdl.org/

1. WinBGIm, a BGI Windows port:
   http://www.cs.colorado.edu/~main/cs1300/doc/bgi/

2. SDL_gfx, graphics drawing primitives and more:
   http://cms.ferzkopp.net/index.php/software/13-sdl-gfx

3. BGI on Wikipedia:
   https://en.wikipedia.org/wiki/Borland_Graphics_Interface

4. Xbgi, a BGI XLib port:
   http://libxbgi.sourceforge.net/

This library is licensed under the Zlib license; please see the file
LICENSE. The 8x8 font comes from A. Schiffler's SDL_gfx.

Brought to you by Guido Gonzato, PhD <guido.gonzato at gmail.com>


ChangeLog
Changelog
---------

v. 2.3.0, 2019-08-01

- added 'auto mode': initgraph() performs automatic screen refresh
  when the environment variable SDL_BGI_RATE is set to 'auto' or to
  the desired refresh rate (e.g. 60). This can be used as an
  alternative to 'slow mode', and it provides a huge performance
  boost. The new function sdlbgiauto() is also available. BEWARE:
  automatic refresh may not work on some NVIDIA graphics cards.
  Patch provided by Marco Diego Aurélio Mesquita.
- added compatibility with C++ programs
- added showerrorbox()
- improved compatibility in bar3d()
- fixed bug in setfillpattern()
- updated Makefiles for Windows (updated to SDL2-2.0.10)
- documentation updates

v. 2.2.4, 2018-11-15
- added some graphics modes for better compatibility
  with old programs
- fixed bug in setvisualpage()
- minor fix in getch()
- fix in putimage() bitwise operations
- modified readimagefile() again
  (reported by Carlos Gustavo Ramirez Rodriguez)
- added test/kaleido.c
- added savestate() and restorestate() in turtle.c
- documentation updates and improvements

v. 2.2.3, 2018-09-18
- fixed bug in setalpha()
- rewritten readimagefile() to change unwanted behaviour
  (reported by Fanourios Attard)

v. 2.2.2, 2018-09-10
- documentation updates (clarification on colours)
- constants TMP_FG_COL, TMP_BG_COL, and TMP_FILL_COL added

v. 2.2.1, 2018-08-08
- renamed getch() to bgi_getch() and added the macro getch()
  to avoid Mingw bug
- delay() rewritten to take care of key presses
- fixed bug in BLUE_VALUE(), GREEN_VALUE(), RED_VALUE(), setalpha()
- added ALPHA_VALUE()
- added test/psychedelia.c
- code revision and cleanup

v. 2.2.0, 2018-07-18
- added setwinoptions() for window title, position, and SDL flags
- added support for multiple windows: extended initwindow(),
  added getcurrentwindow(), setcurrentwindow(), closewindow()
- SDL_QUIT support in getch(), getevent(), kbhit(), xkbhit()
- documentation updates

v. 2.1.2, 2018-06-05
- documentation updates (Code::Blocks, Dev-C++)

v. 2.1.1, 2018-05-25
- fixed SDL_bgi.h (missing stdio.h, stdlib.h, and string.h)
- merged SDL_bgi.h and SDL_bgi_font.h

v. 2.1.0, 2018-01-02
- fixed stupid bug in getch()
- added mouse wheel up/down events
- code cleanup
- improved documentation
- improved turtle graphics
- added CMake support for GNU/Linux
  (thanks to Jani Rautiainen)
- added Mingw-w64 + MSYS2 support
- added Mac OS X support

v. 2.0.8, 2016-11-15
- fixed bug in palette-manipulating functions 
  (reported by Jacek Mainczyk)
- Makefile changes

v. 2.0.7, 2016-11-10
- fixed bug in event()
  (reported by Peje Nilsson)
- added eventtype()

v. 2.0.6, 2016-02-29
- fixed bug in writeimagefile()
  (reported by Manea Alexandru Damian)
- implemented clipping in readimagefile()
- turtle graphics improvements
- documentation improvements

v. 2.0.5, 2015-12-13
- fillpoly() correctly implemented
- demo programs sdlbgidemo.c and fern.c improved
- symbol _SDL_BGI_H defined in SDL_bgi.h
- minor fix in getch()

v. 2.0.4, 2015-11-30
- rewritten to use SDL_Surfaces only instead of
  bitmaps in main RAM
- ellipse() and fillellipse() improved
- xkbhit() added

v. 2.0.3, 2015-09-28
- in "slow mode", screen refresh is also 
  triggered by kbhit() and getch()
- minimal dos.h and conio.h provided
- kbhit() rewritten to avoid buffered keys

v. 2.0.2, 2015-06-03
- fill patterns (not for floodfill() yet)
- setfillpattern()/getfillpattern() implemented
- writeimagefile() added
- documentation updates
- test programs updates

v. 2.0.1, 2015-05-11
- line patterns implemented
- floodfill improvements
- writeimagefile() added
- documentation updates

v. 2.0.0, 2015-04-30
- major release: rewritten to use SDL2 only
- added readimagefile() and logical plotting (XOR, etc.)

v. 1.0.1, 2014-12-09
- Fixed minor bugs in putpixel()
  (reported by Angelo Graziosi)
- changed getimage()/putimage() to include width, heigth info
  in saved bitmap
- changed getimage()/putimage() to include viewport info
- fixed MAXCOLOR value
  (reported by Angelo Graziosi)
- fixed kbhit() behaviour (reported by Angelo Graziosi) - but breaks
  mouseclick() as a result. To be fixed later.

v. 1.0.0, 2014-11-07
- Initial release.


test/dla.c
/* dla.c  -*- C -*-
 * 
 * To compile:
 * gcc -o dla dla.c -lSDL_bgi -lSDL2
 * 
 * Diffusion limited aggregation.
 * Try to run this program with and without 'auto mode',
 * that is setting and unsetting the environment variable
 * SDL_BGI_RATE=auto
 * 
 * By Guido Gonzato, May 2015.
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 * 
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include <graphics.h>

#define SIZE1 800
#define SIZE2 600
#define ALIVE 1
#define DEAD  0

char screen[SIZE1][SIZE2];
void random_walk (void);
unsigned long iterations;

// -----

int main (int argc, char *argv[])
{
  int x, y, stop = 0;
  int gd, gm;
  
  srand (time(NULL));
  
  for (x = 0; x < SIZE1; x++)
    for (y = 0; y < SIZE2; y++)
      screen[x][y] = 0;
  
  for (x = 0; x < 10; x++) 
    screen[random(SIZE1)][random(SIZE2)] = 1; // initial seeds
  
  gd = SDL;
  gm = SDL_800x600;
  initgraph (&gd, &gm, "");
  
  setbkcolor (BLACK);
  cleardevice ();
  iterations = 0;
  
  // set up a palette

  // initialise it to black
  for (x = 0; x < PALETTE_SIZE; x++)
    setrgbpalette (x, 0, 0, 0);
  
  // blue fading to white
  for (x = 0; x < 128; x++)
    setrgbpalette (x, 2*x, 2*x, 255);

  // white fading to yellow
  for (x = 128; x < 256; x++)
    setrgbpalette (x, 255, 255, 255 - x);
  
  // yellow fading to red
  for (x = 256; x < 512; x++)
    setrgbpalette (x, 255, 512 - x, 0);
  
  // red fading to black
  for (x = 512; x < 1024; x += 2)
    setrgbpalette (x, 1024 - 2*x, 0, 0);
  
  // the window is full after ~ 500k iterations.
  
  while (! stop) {
    
    // lay a particle
    random_walk ();
    
    if (iterations % 5000) {
      if (xkbhit ())
	stop = 1;
    }
  }

  printf ("Iterations computed: %lu\n", iterations);
  closegraph ();
  
}

// -----

#define NORTH(n) (n == 0) ? (SIZE2 - 1) : (n - 1)
#define SOUTH(n) (n == SIZE2 - 1) ? (0) : (n + 1)
#define WEST(n)  (n == 0) ? (SIZE1 - 1) : (n - 1)
#define EAST(n)  (n == SIZE1 - 1) ? (0) : (n + 1)

// -----

void random_walk (void)
{
  int x, y, n, done = 0;
  int color;
  
  x = random (SIZE1);
  y = random (SIZE2);
  
  // start random walk until another pixel is hit
  
  do {
    n = random (8);
    
    switch (n) {
     
    case 0:
      x = EAST(x);
      break;
      
    case 1:
      x = EAST(x);
      y = NORTH(y);
      break;
      
    case 2:
      y = NORTH(y);
      break;
            
    case 3:
      x = WEST(x);
      y = NORTH(y);
      break;
      
    case 4:
      x = WEST(x);
      break;
      
    case 5:
      x = WEST(x);
      y = SOUTH(y);
      break;
      
    case 6:
      y = SOUTH(y);
      break;
      
    case 7:
      x = EAST(x);
      y = SOUTH(y);
      break;
    
    } // switch
    
    // any neighbours?
    
    if ( (screen[EAST(x)][y]) ||
	 (screen[EAST(x)][NORTH(y)]) ||
	 (screen[x][NORTH(y)]) ||
	 (screen[WEST(x)][NORTH(y)]) ||
	 (screen[WEST(x)][y]) ||
	 (screen[WEST(x)][SOUTH(y)]) ||
	 (screen[x][SOUTH(y)]) ||
	 (screen[EAST(x)][SOUTH(y)]) ) {
      screen[x][y] = 1;
      iterations++;
      done = 1;
    }
    
  } while (! done);
  
  color = iterations / 150;
  if (color > 1024)
    color = 1024;
  setrgbcolor (color);
  _putpixel (x, y);
  
}

// ----- end of file dla.c


test/dos.h
// dummy dos.h for bgidemo.c

#define far

// ADDED:
int itoa (int val, char* buf, int ignore_me)
{
    const unsigned int radix = 10;

    char* p;
    unsigned int a;        //every digit
    int len;
    char* b;            //start of the digit char
    char temp;
    unsigned int u;

    p = buf;

    if (val < 0)
    {
        *p++ = '-';
        val = 0 - val;
    }
    u = (unsigned int)val;

    b = p;

    do
    {
        a = u % radix;
        u /= radix;

        *p++ = a + '0';

    } while (u > 0);

    len = (int)(p - buf);

    *p-- = 0;

    //swap
    do
    {
        temp = *p;
        *p = *b;
        *b = temp;
        --p;
        ++b;

    } while (b < p);

    return len;
}


test/fern.c
/* fern.c  -*- C -*-
 * 
 * To compile:
 * gcc -o fern fern.c -lSDL_bgi -lSDL2
 * 
 * By Guido Gonzato, May 2015.
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 * 
 */

#include <stdio.h>
#include <unistd.h>
#include <time.h>
#include <math.h>

/* graphics.h could be a link to SDL_bgi.h
 * or another implementation. Let's write this program
 * in a portable way.
 */

#include <graphics.h>

/* ----- */

int main (int argc, char *argv[])
{
#ifndef _SDL_BGI_H
  int gd, gm;
#endif
  int stop = 0;
  long int counter = 0;
  int midx, midy, scale;
  int k = 0, prob;
  float x, y, xx, yy;
  float
    a[4] = {0.0, 0.85, 0.2, -0.15},
    b[4] = {0.0, 0.04, -0.26, 0.28},
    c[4] = {0.0, 0.04, 0.23, 0.26},
    d[4] = {0.16, 0.85, 0.22, 0.24},
    /* e[4] = {0.0, 0.0, 0.0, 0.0}, */
    f[4] = {0.0, 1.6, 1.6, 0.44};
    /* p[4] = {0.01, 0.85, 0.07, 0.07}; */

#ifdef _SDL_BGI_H
  setwinoptions ("This is a fractal fern",
		 SDL_WINDOWPOS_CENTERED,
		 SDL_WINDOWPOS_CENTERED,
		 -1);
  initgraph (NULL, NULL, "");
#else
  initgraph (&gd, &gm, "");
#endif
  srand (time(NULL));
  
#ifdef _SDL_BGI_H
  setbkcolor (COLOR (0, 0, 40));
#else
  setbkcolor (BLACK);
#endif
  cleardevice ();
  setcolor (YELLOW);
  outtextxy (0, 0, "Press a key to exit: ");

  midx = getmaxx () / 2;
  midy = getmaxy () / 2;
  scale = getmaxx () / 16;
  x = y = 0.0;
#ifdef _SDL_BGI_H
  setcolor (COLOR (random (256), random (256), random (256)));
#else
  setcolor (random (WHITE));
#endif
  
  /* kbhit (), mouseclick () and event () are slow. Use them sparingly. */
  
  while (! stop) {
    prob = 1 + random (100);
    if (prob == 1)
      k = 0;
    if ( (prob > 1) && (prob < 87))
      k = 1;
    if ( (prob > 86) && (prob < 94))
      k = 2;
    if (prob > 93)
      k = 3;
    
    /* to use equal probability, just use:
     * k = random (5);
     */
    xx = a[k] * x + b[k] * y; /*  + e[k]; */
    yy = c[k] * x + d[k] * y + f[k];
    x = xx;
    y = yy;
#ifdef _SDL_BGI_H
    _putpixel ((int)(midx + scale * x), (2 * midy - scale * y));
#else
    putpixel ((int)(midx + scale * x), (2 * midy - scale * y), getcolor ());
#endif
    counter++;
    if (0 == counter % 100000) {
#ifdef _SDL_BGI_H
      setcolor (COLOR (random (256), random (256), random (256)));
      refresh ();
#else
      setcolor (random (WHITE));
#endif
      /* let's check for an event only once in a while. */
      if (kbhit ())
	stop = 1;
    }
    if (10000000 == counter) /* 10 million pixels */
      break;
  }
  setcolor (GREEN);
  outtextxy (0, 10, "Ok, leaving in 2 seconds");
  delay (2000);
  closegraph ();
  return 0;
}

/* --  end of file fern.c --- */


build.sh
#!/bin/sh

VERSION=$(cat VERSION)

mkdir -p build && cd build/
cmake -DCMAKE_INSTALL_PREFIX=/usr ..
PACKAGE_DONE=0

# is this system .deb based?
DEB=$(which dpkg 2> /dev/null)
if [ $? = 0 ] ; then
  DEB_ARCH=$(dpkg --print-architecture)
  printf "\n*** This system looks Debian-based.\n\n"
  cpack -G "DEB" ..
  DEB_NAME="sdl_bgi_"$VERSION"-1_"$DEB_ARCH".deb"
  /bin/mv *deb $DEB_NAME
  printf "\n*** Package $DEB_NAME created in subdirectory build/.\n\n"
  PACKAGE_DONE=1
fi

# is this system .rpm based?
RPM=$(which rpm 2> /dev/null)
if [ $? = 0 ] ; then
  RPM_ARCH=$(uname -m)
  printf "\n*** This system looks RPM-based.\n\n"
  cpack -G "RPM" ..
  RPM_NAME="SDL_bgi-"$VERSION"-1."$RPM_ARCH".rpm"
  /bin/mv *rpm $RPM_NAME
  printf "\n*** Package $RPM_NAME created in subdirectory build/.\n\n"
  PACKAGE_DONE=1
fi

if [ $PACKAGE_DONE = 0 ] ; then
  printf "\n*** I don't know what package to build.\n\n"
fi


src/SDL_bgi.h
// SDL_bgi.h	-*- C -*-

// A BGI (Borland Graphics Library) implementation based on SDL2.
// Easy to use, pretty fast, and useful for porting old programs.
// Guido Gonzato, PhD
// August 1, 2019

/*
This software is provided 'as-is', without any express or implied
warranty.  In no event will the authors be held liable for any damages
arising from the use of this software.

Permission is granted to anyone to use this software for any purpose,
including commercial applications, and to alter it and redistribute it
freely, subject to the following restrictions:

1. The origin of this software must not be misrepresented; you must not
   claim that you wrote the original software. If you use this software
   in a product, an acknowledgment in the product documentation would be
   appreciated but is not required.
2. Altered source versions must be plainly marked as such, and must not be
   misrepresented as being the original software.
3. This notice may not be removed or altered from any source distribution.
*/

#ifndef _SDL_BGI_H
#define _SDL_BGI_H

// SDL2 stuff
#include <SDL2/SDL.h>
#include <SDL2/SDL_keycode.h>

#include <stdio.h>   // for fprintf()
#include <stdlib.h>  // for exit(), calloc()
#include <math.h>    // for sin(), cos()
#include <string.h>  // for strlen(), memcpy()

#define SDL_BGI_VERSION 2.3.0

enum { NOPE, YEAH };
#define BGI_WINTITLE_LEN 512 // more than enough

// number of concurrent windows that can be created

#define NUM_BGI_WIN 16

// everything gets drawn here

extern SDL_Window   *bgi_window;
extern SDL_Renderer *bgi_renderer;
extern SDL_Texture  *bgi_texture;

// available visual pages

#define VPAGES 4

// BGI fonts

// only DEFAULT_FONT (8x8) is implemented
enum {
  DEFAULT_FONT, TRIPLEX_FONT, SMALL_FONT, SANSSERIF_FONT,
  GOTHIC_FONT, BIG_FONT, SCRIPT_FONT, SIMPLEX_FONT,
  TRIPLEX_SCR_FONT, COMPLEX_FONT, EUROPEAN_FONT, BOLD_FONT
};

enum { HORIZ_DIR, VERT_DIR };

#define USER_CHAR_SIZE 0

enum {
  LEFT_TEXT, CENTER_TEXT, RIGHT_TEXT,
  BOTTOM_TEXT = 0, TOP_TEXT = 2
};

// BGI colours

enum {
  BLACK, BLUE, GREEN, CYAN, RED, MAGENTA, BROWN,
  LIGHTGRAY, DARKGRAY, LIGHTBLUE, LIGHTGREEN, LIGHTCYAN,
  LIGHTRED, LIGHTMAGENTA, YELLOW, WHITE, MAXCOLORS = 15
};

// temporary colours

enum { TMP_FG_COL = 16, TMP_BG_COL = 17, TMP_FILL_COL = 18 };

// line style, thickness, and drawing mode

enum { NORM_WIDTH = 1, THICK_WIDTH = 3 };

enum { SOLID_LINE, DOTTED_LINE, CENTER_LINE, DASHED_LINE, USERBIT_LINE };

enum { COPY_PUT, XOR_PUT, OR_PUT, AND_PUT, NOT_PUT };

// fill styles

enum {
  EMPTY_FILL, SOLID_FILL, LINE_FILL, LTSLASH_FILL, SLASH_FILL,
  BKSLASH_FILL, LTBKSLASH_FILL, HATCH_FILL, XHATCH_FILL,
  INTERLEAVE_FILL, WIDE_DOT_FILL, CLOSE_DOT_FILL, USER_FILL
};

// mouse buttons

#define WM_LBUTTONDOWN  SDL_BUTTON_LEFT
#define WM_MBUTTONDOWN  SDL_BUTTON_MIDDLE
#define WM_RBUTTONDOWN  SDL_BUTTON_RIGHT
#define WM_WHEEL        SDL_MOUSEWHEEL
#define WM_WHEELUP      SDL_USEREVENT
#define WM_WHEELDOWN    SDL_USEREVENT + 1
#define WM_MOUSEMOVE    SDL_MOUSEMOTION

#define PALETTE_SIZE    4096

#define KEY_HOME        SDLK_HOME
#define KEY_LEFT        SDLK_LEFT
#define KEY_UP          SDLK_UP
#define KEY_RIGHT       SDLK_RIGHT
#define KEY_DOWN        SDLK_DOWN
#define KEY_PGUP        SDLK_PAGEUP
#define KEY_PGDN        SDLK_PAGEDOWN
#define KEY_END         SDLK_END
#define KEY_INSERT      SDLK_INSERT
#define KEY_DELETE      SDLK_DELETE
#define KEY_F1          SDLK_F1
#define KEY_F2          SDLK_F2
#define KEY_F3          SDLK_F3
#define KEY_F4          SDLK_F4
#define KEY_F5          SDLK_F5
#define KEY_F6          SDLK_F6
#define KEY_F7          SDLK_F7
#define KEY_F8          SDLK_F8
#define KEY_F9          SDLK_F9
#define KEY_F10         SDLK_F10
#define KEY_F11         SDLK_F11
#define KEY_F12         SDLK_F12
#define KEY_CAPSLOCK    SDLK_CAPSLOCK
#define KEY_LEFT_CTRL   SDLK_LCTRL
#define KEY_RIGHT_CTRL  SDLK_RCTRL
#define KEY_LEFT_SHIFT  SDLK_LSHIFT
#define KEY_RIGHT_SHIFT SDLK_RSHIFT
#define KEY_LEFT_ALT    SDLK_LALT
#define KEY_RIGHT_ALT   SDLK_RALT
#define KEY_ALT_GR      SDLK_MODE
#define KEY_LGUI        SDLK_LGUI
#define KEY_RGUI        SDLK_RGUI
#define KEY_MENU        SDLK_MENU
#define KEY_TAB         SDLK_TAB
#define KEY_BS          SDLK_BACKSPACE
#define KEY_RET         SDLK_RETURN
#define KEY_PAUSE       SDLK_PAUSE
#define KEY_SCR_LOCK    SDLK_SCROLLOCK
#define KEY_ESC         SDLK_ESCAPE

#define QUIT            SDL_QUIT

// graphics modes. Expanded from the original GRAPHICS.H

enum {
  DETECT = -1,
  grOk = 0, SDL = 0,
  // all modes @ 320x200 
  SDL_320x200 = 1, SDL_CGALO = 1, CGA = 1, CGAC0 = 1, CGAC1 = 1,
  CGAC2 = 1, CGAC3 = 1, MCGAC0 = 1, MCGAC1 = 1, MCGAC2 = 1,
  MCGAC3 = 1, ATT400C0 = 1, ATT400C1 = 1, ATT400C2 = 1, ATT400C3 = 1,
  // all modes @ 640x200
  SDL_640x200 = 2, SDL_CGAHI = 2, CGAHI = 2, MCGAMED = 2,
  EGALO = 2, EGA64LO = 2,
  // all modes @ 640x350
  SDL_640x350 = 3, SDL_EGA = 3, EGA = 3, EGAHI = 3,
  EGA64HI = 3, EGAMONOHI = 3,
  // all modes @ 640x480
  SDL_640x480 = 4, SDL_VGA = 4, VGA = 4, MCGAHI = 4, VGAHI = 4,
  IBM8514LO = 4,
  // all modes @ 720x348
  SDL_720x348 = 5, SDL_HERC = 5,
  // all modes @ 720x350
  SDL_720x350 = 6, SDL_PC3270 = 6, HERCMONOHI = 6,
  // all modes @ 800x600
  SDL_800x600 = 7, SDL_SVGALO = 7, SVGA = 7,
  // all modes @ 1024x768
  SDL_1024x768 = 8, SDL_SVGAMED1 = 8,
  // all modes @ 1152x900
  SDL_1152x900 = 9, SDL_SVGAMED2 = 9,
  // all modes @ 1280x1024
  SDL_1280x1024 = 10, SDL_SVGAHI = 10,
  // all modes @ 1366x768
  SDL_1366x768 = 11, SDL_WXGA = 11,
  // other
  SDL_USER = 12, SDL_FULLSCREEN = 13
};

// libXbgi compatibility

#define X11_CGALO       SDL_CGALO
#define X11_CGAHI       SDL_CGAHI
#define X11_EGA         SDL_EGA
#define X11             SDL
#define X11_VGA         SDL_VGA
#define X11_640x480     SDL_640x480
#define X11_HERC        SDL_HERC
#define X11_PC3270      SDL_PC3270
#define X11_SVGALO      SDL_SVGALO
#define X11_800x600     SDL_800x600
#define X11_SVGAMED1    SDL_SVGAMED1
#define X11_1024x768    SDL_1024x768
#define X11_SVGAMED2    SDL_SVGAMED2
#define X11_1152x900    SDL_1152x900
#define X11_SVGAHI      SDL_SVGAHI
#define X11_1280x1024   SDL_1280x1024
#define X11_WXGA        SDL_WXGA
#define X11_1366x768    SDL_1366x768
#define X11_USER        SDL_USER
#define X11_FULLSCREEN  SDL_FULLSCREEN

// structs

struct arccoordstype {
  int x;
  int y;
  int xstart;
  int ystart;
  int xend;
  int yend;
};

struct date {
  int da_year;
  int da_day;
  int da_mon;
};

struct fillsettingstype {
  int pattern;
  int color;
};

struct linesettingstype {
  int linestyle;
  unsigned int upattern;
  int thickness;
};

struct palettetype {
  unsigned char size;
  signed char colors[MAXCOLORS + 1];
};

struct textsettingstype {
  int font;
  int direction;
  int charsize;
  int horiz;
  int vert;
};

struct viewporttype {
  int left;
  int top;
  int right;
  int bottom;
  int clip;
};

// prototypes - standard BGI
// make them C++ compatible

#ifdef __cplusplus
extern "C" {
#endif

void arc (int, int, int, int, int);
void bar3d (int, int, int, int, int, int);
void bar (int, int, int, int);
void circle (int, int, int);
void cleardevice ();
void clearviewport ();
void closegraph (void);
void delay (int msec);
void detectgraph (int *, int *);
void drawpoly (int, int *);
void ellipse (int, int, int, int, int, int);
void fillellipse (int, int, int, int);
void fillpoly (int, int *);
void floodfill (int, int, int);
int  getactivepage (void);
void getarccoords (struct arccoordstype *);
void getaspectratio (int *, int *);
int  getbkcolor (void);
int  bgi_getch (void);
// circumvents Mingw bug
#define getch bgi_getch
int  getcolor (void);
struct palettetype *getdefaultpalette (void);
char *getdrivername (void);
void getfillpattern (char *);
void getfillsettings (struct fillsettingstype *);
int  getgraphmode (void);
void getimage (int, int, int, int, void *);
void getlinesettings (struct linesettingstype *);
int  getmaxcolor (void);
int  getmaxmode (void);
int  getmaxx (void);
int  getmaxy (void);
char *getmodename (int);
void getmoderange (int, int *, int *);
void getpalette (struct palettetype *);
int  getpalettesize (struct palettetype *);
unsigned int getpixel (int, int);
void gettextsettings (struct textsettingstype *);
void getviewsettings (struct viewporttype *);
int  getvisualpage (void);
int  getx (void);
int  gety (void);
void graphdefaults ();
char *grapherrormsg (int);
int  graphresult (void);
unsigned imagesize (int, int, int, int);
void initgraph (int *, int *, char *);
int  installuserdriver (char *name, int (*detect)(void));
int  installuserfont (char *);
int  kbhit (void);
void line (int, int, int, int);
void linerel (int dx, int dy);
void lineto (int x, int y);
void moverel (int, int);
void moveto (int, int);
void outtext (char *);
void outtextxy (int, int, char *);
void pieslice (int, int, int, int, int);
void putimage (int, int, void *, int);
void putpixel (int, int, int);
#define random(range) (rand() % (range))
void readimagefile (char *, int, int, int, int);
void rectangle (int, int, int, int);
int  registerbgidriver (void (*driver)(void));
int  registerbgifont (void (*font)(void));
void restorecrtmode (void);
void sector (int, int, int, int, int, int);
void setactivepage (int);
void setallpalette (struct palettetype *);
void setaspectratio (int, int);
void setbkcolor (int);
void setcolor (int);
void setfillpattern (char *, int);
void setfillstyle (int, int);
unsigned setgraphbufsize (unsigned);
void setgraphmode (int);
void setlinestyle (int, unsigned, int);
void setpalette (int, int);
void settextjustify (int, int);
void settextstyle (int, int, int);
void setusercharsize (int, int, int, int);
void setviewport (int, int, int, int, int);
void setvisualpage (int);
void setwritemode (int);
int  textheight (char *);
int  textwidth (char *);
void writeimagefile (char *, int, int, int, int);

// SDL_bgi extensions

int  ALPHA_VALUE (int);
int  BLUE_VALUE (int);
void closewindow (int);
int  COLOR (int, int, int);
int  event (void);
int eventtype (void);
void freeimage (void *);
int  getcurrentwindow (void);
int  getevent (void);
void getmouseclick (int, int *, int *);
int  GREEN_VALUE (int);
void initwindow (int, int);
int  IS_BGI_COLOR (int color);
int  ismouseclick (int);
int  IS_RGB_COLOR (int color);
int  mouseclick (void);
int  mousex (void);
int  mousey (void);
void _putpixel (int, int);
int  RED_VALUE (int );
void refresh (void);
void sdlbgiauto (void);
void sdlbgifast (void);
void sdlbgislow (void);
void setalpha (int, Uint8);
void setbkrgbcolor (int);
void setblendmode (int);
void setcurrentwindow (int);
void setrgbcolor (int);
void setrgbpalette (int, int, int, int);
void setwinoptions (char *, int, int, Uint32);
void showerrorbox (const char *);
void swapbuffers (void);
int  xkbhit (void);

#ifdef __cplusplus
}
#endif

// #ifndef _SDL_BGI_H

#endif // _SDL_BGI_H

// --- End of file SDL_bgi.h
